003_bandwidth_check/bandwidth_check.cu

```diff
@@ -1,49 +1,128 @@
 #include <cstdio>
+#include <cstdlib>
+#include <getopt.h>
 #include <cuda_runtime.h>
-constexpr int NUM_COPY_ITERATIONS = 1000;
+#include <vector>
 
-void checkBandwidth(cudaStream_t& stream, size_t dataSize) {
-    // Allocate and load host memory.
-    float* hData = new float[dataSize];
+// Default values
+constexpr int DEFAULT_NUM_COPY_ITERATIONS = 1000;
+constexpr int DEFAULT_NUM_STREAMS = 4;
+constexpr int DEFAULT_WARMUP_ITERATIONS = 10;
+constexpr size_t DEFAULT_DATA_SIZE = 1 << 24; // 16M floats
+
+void checkBandwidth(size_t dataSize, int numCopyIterations, int numStreams, int warmupIterations) {
+    // Allocate pinned host memory.
+    float* hData;
+    ::cudaHostAlloc(&hData, dataSize * sizeof(float), cudaHostAllocDefault);
+
+    // Populate the host array with values.
     for (size_t i = 0; i < dataSize; ++i) {
         hData[i] = static_cast<float>(i);
     }
+
     // Allocate device memory.
     float* dData;
     ::cudaMalloc(&dData, dataSize * sizeof(float));
-    ::cudaStreamSynchronize(stream);
+
     // Create CUDA events for timing purposes.
     cudaEvent_t start, stop;
     ::cudaEventCreate(&start);
     ::cudaEventCreate(&stop);
+
+    // Create multiple CUDA streams.
+    std::vector<cudaStream_t> streams(numStreams);
+    for (int i = 0; i < numStreams; ++i) {
+        ::cudaStreamCreate(&streams[i]);
```

Comment:
I don't see why this would make a difference. Did you try it on a single stream?